File: apps/ios/Flashcards/Flashcards/Settings/SettingsAttentionSupport.swift
import Foundation

enum SettingsAttentionIssue: Hashable, Sendable {
    case accountNotLinked
}

enum SettingsAttentionTarget: Hashable, Sendable {
    case settingsTab
    case accountStatusRow
    case accountStatusPrimaryAction
}

struct SettingsAttentionSummary: Equatable, Sendable {
    let settingsTabCount: Int
    let accountStatusRowCount: Int
    let accountStatusPrimaryActionCount: Int
}

func makeSettingsAttentionIssues(cloudState: CloudAccountState?) -> [SettingsAttentionIssue] {
    switch cloudState {
    case .linked:
        return []
    case nil, .disconnected, .linkingReady, .guest:
        return [.accountNotLinked]
    }
}

func makeSettingsAttentionSummary(issues: [SettingsAttentionIssue]) -> SettingsAttentionSummary {
    SettingsAttentionSummary(
        settingsTabCount: makeSettingsAttentionTargetCount(
            issues: issues,
            target: .settingsTab
        ),
        accountStatusRowCount: makeSettingsAttentionTargetCount(
            issues: issues,
            target: .accountStatusRow
        ),
        accountStatusPrimaryActionCount: makeSettingsAttentionTargetCount(
            issues: issues,
            target: .accountStatusPrimaryAction
        )
    )
}

private func makeSettingsAttentionTargetCount(
    issues: [SettingsAttentionIssue],
    target: SettingsAttentionTarget
) -> Int {
    issues.reduce(0) { count, issue in
        count + (settingsAttentionTargets(issue: issue).contains(target) ? 1 : 0)
    }
}

private func settingsAttentionTargets(issue: SettingsAttentionIssue) -> [SettingsAttentionTarget] {
    switch issue {
    case .accountNotLinked:
        return [.settingsTab, .accountStatusRow, .accountStatusPrimaryAction]
    }
}

File: apps/android/feature/settings/src/main/java/com/flashcardsopensourceapp/feature/settings/SettingsAttentionSupport.kt
package com.flashcardsopensourceapp.feature.settings

import com.flashcardsopensourceapp.data.local.model.cloud.CloudAccountState

enum class SettingsAttentionIssue {
    ACCOUNT_NOT_LINKED
}

data class SettingsAttentionSummary(
    val settingsTabCount: Int,
    val accountStatusRowCount: Int,
    val accountStatusPrimaryActionCount: Int
)

fun makeSettingsAttentionIssues(cloudState: CloudAccountState): List<SettingsAttentionIssue> {
    return when (cloudState) {
        CloudAccountState.DISCONNECTED,
        CloudAccountState.LINKING_READY,
        CloudAccountState.GUEST -> listOf(SettingsAttentionIssue.ACCOUNT_NOT_LINKED)
        CloudAccountState.LINKED -> emptyList()
    }
}

fun makeSettingsAttentionSummary(issues: List<SettingsAttentionIssue>): SettingsAttentionSummary {
    val accountNotLinkedCount: Int = issues.count { issue ->
        issue == SettingsAttentionIssue.ACCOUNT_NOT_LINKED
    }

    return SettingsAttentionSummary(
        settingsTabCount = issues.size,
        accountStatusRowCount = accountNotLinkedCount,
        accountStatusPrimaryActionCount = accountNotLinkedCount
    )
}
